include/experimental/__p1673_bits/blas3_triangular_matrix_matrix_solve.hpp

```diff
@@ -188,35 +167,29 @@ void trsm_upper_triangular_right_side(
   }
 }
 
-template<class ElementType_A,
-         class SizeType_A, ::std::size_t numRows_A, ::std::size_t numCols_A,
-         class Layout_A,
-         class Accessor_A,
-         class DiagonalStorage,
-         class ElementType_B,
-         class SizeType_B, ::std::size_t numRows_B, ::std::size_t numCols_B,
-         class Layout_B,
-         class Accessor_B,
-         class ElementType_X,
-         class SizeType_X, ::std::size_t numRows_X, ::std::size_t numCols_X,
-         class Layout_X,
-         class Accessor_X>
+template<
+  P1673_MATRIX_TEMPLATE_PARAMETERS( A ),
+  class DiagonalStorage,
+  P1673_MATRIX_TEMPLATE_PARAMETERS( B ),
+  P1673_MATRIX_TEMPLATE_PARAMETERS( X )
+>
 void trsm_lower_triangular_right_side(
-  std::experimental::mdspan<ElementType_A, std::experimental::extents<SizeType_A, numRows_A, numCols_A>, Layout_A, Accessor_A> A,
+  P1673_MATRIX_PARAMETER( A ),
   DiagonalStorage d,
-  std::experimental::mdspan<ElementType_B, std::experimental::extents<SizeType_B, numRows_B, numCols_B>, Layout_B, Accessor_B> B,
-  std::experimental::mdspan<ElementType_X, std::experimental::extents<SizeType_X, numRows_X, numCols_X>, Layout_X, Accessor_X> X)
+  P1673_MATRIX_PARAMETER( B ),
+  P1673_MATRIX_PARAMETER( X ))
 {
   constexpr bool explicit_diagonal =
     std::is_same_v<DiagonalStorage, explicit_diagonal_t>;
   using size_type = ::std::common_type_t<SizeType_A, SizeType_B, SizeType_X>;
+  using signed_index_type = ::std::make_signed_t<size_type>;
```

Comment:
Question: is this preferred over `ptrdiff_t` and will/should the usage of `ptrdiff_t` for index decrement at other places be replaced with this?

Reply:
@youyu3 Thanks for asking!  The two priorities for the sequential / "fall-back" implementation are

1. avoid unwarranted overflow; and
2. avoid build warnings (e.g., on signed-to-unsigned conversion).

A reasonable implementation approach would be to use `size_t` if the common type of all the index types is unsigned, and `ptrdiff_t` if the common type is signed.  It doesn't matter, as long as we use an index type that can't overflow.